src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -367,7 +367,7 @@ CosineTree::~CosineTree()
     delete right;
 }
 
-void CosineTree::ModifiedGramSchmidt(CosineNodeQueue& treeQueue,
+void CosineTree::ModifiedGramSchmidt(CosineNodeQueue treeQueue,
```

Comment:
I think we should keep this as a reference---I believe the method is meant to modify `treeQueue` in-place. :+1:

Reply:
Okay, I will check it. Also, do you think making copies is okay?

Reply:
In general we should try to avoid making copies wherever possible. :+1:

Reply:
[This](https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/tree/cosine_tree/cosine_tree.cpp#L370) is the function you mentioned. I don't think we are modifying `treeQueue` anywhere here.

I might be missing something, can you point at the exact line where we are modifying `treeQueue`

Reply:
When you pass a reference to an object, it does not copy the object.  When you pass an object itself, it will be copied.  So, in the example here, if we pass a `CosineNodeQueue&`, it will not be copied---but if we pass just a `CosineNodeQueue`, then it will be copied.

In this particular case, we are modifying elements that are inside `treeQueue`---therefore we need to pass a reference for it to work properly. :+1: